### PR TITLE
Added config option to set size of transcode cache and cadence to enf…

### PR DESCRIPTION
**Pull request by brian-doherty**
…orce that sizing via ejection.

### CODE REVIEW

**Comment by brian-doherty** [comment]
Resolves #501

**Comment by brian-doherty** [comment]
@sentriz Would love to get your feedback on this when you get a minute. Thanks.

**Comment by sentriz** [comment]
hiya Brian sorry for the delay . will hopefully take a look shortly

**Comment by zpuskas** [comment]
@brian-doherty Could you please also update contrib/config accordingly with this new option?

**Comment by brian-doherty** [comment]
> @brian-doherty Could you please also update [contrib/config](https://github.com/sentriz/gonic/blob/master/contrib/config) accordingly with this new option?

Done. Lint and test are running now but will pass.

**Comment by brian-doherty** [comment]
@sentriz I know you're busy but I would love you to have a look at this PR when you get a sec.

File: transcode/transcoder_caching.go [hunk 1]
```diff
@@ -64,6 +68,41 @@ func (t *CachingTranscoder) Transcode(ctx context.Context, profile Profile, in s
...
+
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].info.ModTime().Before(files[j].info.ModTime())
+	})
+
+	for total > int64(t.limitMb)*1024*1024 {
+		curFile := files[0]
+		files = files[1:]
+		total -= curFile.info.Size()
+		unlock := t.locks.Lock(curFile.path)
```
Comment (sentriz) [inline transcode/transcoder_caching.go:100]: the t.locks is a keyedMutex but the keys are profile strings (ffmpeg commands with args and paths) not paths. so this has little effect i think

maybe we should have a global clean lock which the normal transcode thing respects too?

something like

```
type CachingTranscoder struct {
	...
	locks      keyedMutex
    cleanLock  sync.RWMutex
}

...

func (t *CachingTranscoder) Transcode(...) error {
    ...

    t.cleanLock.RLock()
    defer t.cleanLock.RUnlock()
    
    ...

	key := cacheKey(name, args)
	unlock := t.locks.Lock(key)
	defer unlock()
}

func (t *CachingTranscoder) CacheEject() error {
    t.cleanLock.Lock()
    defer t.cleanLock.Unlock()
}
```

that way the cleaning process locks the entire cache, while multiple goroutines looking to read the cache work fine, just not while we're cleaning

wdyt?

**Review by sentriz — COMMENTED** [review]
looks good! just a couple things

File: transcode/transcoder_caching.go [hunk 2]
```diff
@@ -7,21 +7,24 @@ import (
...
 	transcoder Transcoder
+	limitMb    int
 	locks      keyedMutex
 }
 
 var _ Transcoder = (*CachingTranscoder)(nil)
 
-func NewCachingTranscoder(t Transcoder, cachePath string) *CachingTranscoder {
-	return &CachingTranscoder{transcoder: t, cachePath: cachePath}
+func NewCachingTranscoder(t Transcoder, cachePath string, limitMb int) *CachingTranscoder {
```
Comment (sentriz) [inline transcode/transcoder_caching.go:26]: looks megabits to me. how about a big B for bytes

```suggestion
func NewCachingTranscoder(t Transcoder, cachePath string, limitMB int) *CachingTranscoder {
```

File: transcode/transcoder_caching.go [hunk 3]
```diff
@@ -64,6 +68,41 @@ func (t *CachingTranscoder) Transcode(ctx context.Context, profile Profile, in s
 	return nil
 }
 
+func (t *CachingTranscoder) CacheEject() {
```
[the remark this answers is not in the document]
Comment (brian-doherty) [reply transcode/transcoder_caching.go]: Done.

File: transcode/transcoder_caching.go [hunk 4]
```diff
@@ -64,6 +68,41 @@ func (t *CachingTranscoder) Transcode(ctx context.Context, profile Profile, in s
...
+	// Delete LRU cache files that exceed size limit. Use last modified time.
+	type file struct {
+		path string
+		info os.FileInfo
+	}
+
+	var files []file
+	var total int64 = 0
+
+	_ = filepath.Walk(t.cachePath, func(path string, info os.FileInfo, err error) error {
```
[the remark this answers is not in the document]
Comment (brian-doherty) [reply transcode/transcoder_caching.go]: Done.

File: transcode/transcoder_caching.go (hunk 1 above)
Comment (brian-doherty) [reply transcode/transcoder_caching.go]: Good call and my apologies. Done.

File: transcode/transcoder_caching.go (hunk 2 above)
Comment (brian-doherty) [reply transcode/transcoder_caching.go]: Done.

**Comment by brian-doherty** [comment]
@sentriz All comments addressed. See what you think and thanks!

File: cmd/gonic/gonic
Comment (sentriz) [inline cmd/gonic/gonic:1]: looks like this was committed accidentally

Comment (brian-doherty) [reply cmd/gonic/gonic:1]: My bad. Fixed.

**Comment by sentriz** [comment]
looks great, thanks! sorry for the delay

**Pull request merged**